### DISCUSSION TITLE
Remove linters/tests from Netlify build?

### DISCUSSION THREAD

**Discussion by phritolay** [ample/gatsby-starter-ample#196] [Ideas]
We [added](https://github.com/ample/gatsby-starter-ample/issues/168) a `pre-push` hook using [husky](https://www.npmjs.com/package/husky) to run specs prior to opening a PR to help a dev know if tests are failing sooner, rather than waiting for a build to fail on Netlify. Currently linters and tests run with both a `pre-push` hook and on Netlify builds, so it would be nice to remove the scripts from `netlify:build` to speed up builds. However the `pre-push` hook can be [bypassed using `--no-verify`](https://stackoverflow.com/a/29545973/2241124) which could result in failing specs and a passing build. How should we handle this? My initial thought is we could run specs using GitHub Actions to ensure the specs are not bypassed by a dev, but will help speed up the build process.

**Comment by dholloran-ample** [comment]
@phritolay Would the Github Action block a PR from being merged until they pass? There is a small chance that the pre-commit hook would be bypassed but it still exists.

**Side note:** Would be nice to stop deploy for `main`/`development` if possible but I'm more worried about the PRs. 
    - I wonder if there is a way to only run the tests on Netlify for deployment branches and not previews. Netlify has a bunch of env variables this may be able to be scripted somehow (**I'd be happy to investigate if we feel this is worth it**)

**Comment by seancdavis** [comment]
I think @dholloran-ample is on to something. We may be overthinking this when [deploy contexts](https://www.netlify.com/blog/2016/08/30/introducing-deploy-contexts-in-netlify/) could make this simple. Run specs and linting only in deploy previews and not for production builds?

Is it overkill to have the pre-push hook _and_ have Netlify run the specs/linting?

There's something really nice about having those things run locally on push. They aren't long-running processes, but it's really annoying when a build fails in a PR for something that was quick to fix.

**Comment by seancdavis** [comment]
Based on our Dev Lab conversation this week, we'd like to:

1. Keep the pre-push hook in place to protect against wasting time running builds that are going to fail.
2. Allow Netlify to run the entire CI pipeline. Consider moving the linters and test suite into a Netlify plugin and run pre-build. (As the plugin ecosystem matures, there's a lot we could do with this process.)
3. Explore being able to skip the linting and specs based on where the build originated from.

@phritolay @dholloran-ample Do you agree with this direction?

**Comment by dholloran-ample** [comment]
@seancdavis - Yes, that seems good to me.
